dali/kernels/slice/slice_flip_normalize_permute_pad_cuda_impl.cuh

```diff
@@ -31,122 +32,153 @@
+__device__ void SliceFlipNormalizePermutePadFunc(
+    Out *__restrict__ out, const In *__restrict__ in,
+    const fast_div<uint64_t> *out_strides, const int64_t *in_strides, const int64_t *out_shape,
+    const int64_t *in_shape, const int64_t *anchor, const Out *__restrict__ fill_values,
+    const float *__restrict__ norm_add, const float *__restrict__ norm_mul, int channel_dim,
+    uint64_t offset, uint64_t block_end) {
+  if (Dims > 1 && out_strides[Dims - 1] == in_strides[Dims - 1] && anchor[Dims - 1] == 0 &&
+      out_shape[Dims - 1] == in_shape[Dims - 1] && channel_dim != Dims - 1 &&
+      (!NeedFlip || (in_strides[Dims - 1] < 0) == (in_strides[Dims - 2] < 0))) {
+    const int NextDims = Dims > 1 ? Dims - 1 : 1;
+    SliceFlipNormalizePermutePadFunc<NeedFlip, NeedNormalize, NeedPad, NextDims, Out, In, false>(
+        out, in, out_strides, in_strides, out_shape, in_shape, anchor, fill_values, norm_add,
+        norm_mul, channel_dim, offset, block_end);
     return;
   }
 
-  const bool innermost_is_dense = (out_strides[Dims-1] == 1);
+  constexpr int LastDim = Dims - 1;
+  int64_t inner_in_anchor = anchor[LastDim];
+  int64_t inner_in_extent = in_shape[LastDim];
+  int64_t inner_out_extent = out_shape[LastDim];
+  if (!AllDims) {  // if we fused dimensions, adjust inner dimension's anchor and extent
+    inner_in_anchor = anchor[LastDim] * in_strides[LastDim];
+    inner_in_extent = in_shape[LastDim] * in_strides[LastDim];
+    inner_out_extent = out_shape[LastDim] * in_strides[LastDim];
+  }
+
   for (; offset < block_end; offset += blockDim.x) {
-    size_t idx = offset;
-    size_t out_idx = offset;
-    size_t in_idx = 0;
-    unsigned norm_i = 0;
-    bool pad = false;
-
-    for (unsigned d = 0; d < Dims; d++) {
-      unsigned out_stride = static_cast<unsigned>(out_strides[d]);
-      unsigned i_d;
-      if (d == Dims-1 && innermost_is_dense) {
-        i_d = idx;
-        idx = 0;
-      } else {
-        i_d = idx / out_stride;
-        idx %= out_stride;
+    uint64_t idx = offset;
+    uint64_t out_idx = idx;
+
+    // We can avoid division in the last dimension because know the strides are 1
+    // (or we treat them as 1 if we fused dimensions)
+    int i_c = 0;
+    int i_d;
+    bool out_of_bounds = false;
+    uint64_t in_idx = 0;
+
+    #pragma unroll
+    for (int d = 0; d < Dims - 1; d++) {
+      i_d = div_mod(idx, idx, out_strides[d]);
+      if (d == channel_dim)
+        i_c = i_d;
+      if (NeedPad) {
+        auto in_i_d = NeedFlip && in_strides[d] < 0 ? anchor[d] + out_shape[d] - 1 - i_d
+                                                    : anchor[d] + i_d;
+        out_of_bounds |= is_out_of_bounds(in_i_d, in_shape[d]);
       }
-      if (pad = (should_pad && i_d >= out_shape[d]))
-        break;
+      in_idx += i_d * in_strides[d];
+    }
 
-      if (d == norm_dim)
-        norm_i = i_d;
+    constexpr int d = Dims - 1;
+    i_d = idx;  // out_strides[d] is treated as 1
+    idx = 0;
```

Comment:
I don't see `idx` used anywhere after that line.

Reply:
yeah I know, it was to show that this is equivalent to `i_d = div_mod(idx, idx, out_strides[d]);`. That being said, I think I can remove it